[Form Control Refresh] Focus ring clipped for native <select> buttons

lilyspiniolas opened the pull request:
#### 77ba7337d7cdd989edca5358355aeaf95a2d5f06
<pre>
[Form Control Refresh] Focus ring clipped for native &lt;select&gt; buttons
<a href="https://bugs.webkit.org/show_bug.cgi?id=295256">https://bugs.webkit.org/show_bug.cgi?id=295256</a>
<a href="https://rdar.apple.com/154729412">rdar://154729412</a>

Reviewed by Aditya Keerthi.

After 296129@main, native select buttons no longer had their border
radius reset, causing bleed avoidance to apply which resulted in the
focus ring getting clipped.

With these changes, bleed avoidance (except when handled directly by
the theme) no longer applies to any painted control.

* LayoutTests/fast/forms/select/select-focus-ring-is-not-clipped-expected-mismatch.html: Added.
* LayoutTests/fast/forms/select/select-focus-ring-is-not-clipped.html: Added.
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::determineBleedAvoidance const):
* Source/WebCore/rendering/RenderTheme.h:
(WebCore::RenderTheme::mayNeedBleedAvoidance const):
* Source/WebCore/rendering/cocoa/RenderThemeCocoa.h:
* Source/WebCore/rendering/cocoa/RenderThemeCocoa.mm:
(WebCore::RenderThemeCocoa::mayNeedBleedAvoidance const):

Canonical link: <a href="https://commits.webkit.org/296951@main">https://commits.webkit.org/296951@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f8f600e9be6df79df4e339675ec5de72806eae74

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/110053 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/29712 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/20144 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/116075 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/60301 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/112016 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/30390 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/38299 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/83671 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/113001 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/24236 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/99110 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64115 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/23604 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/17250 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/59871 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/93616 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/17301 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/118866 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/37093 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/27497 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/92640 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/37465 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/95378 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/92466 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23571 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/37454 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/15192 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/32975 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/36987 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/42458 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/36649 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/39989 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/38358 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on LayoutTests/fast/forms/select/select-focus-ring-is-not-clipped-expected-mismatch.html:
```diff
@@ -0,0 +1,27 @@
+        margin: 0;
+        padding: 0;
+        padding-top: 5px;
+	 }
```
Indentation is off.

pxlcoder commented on LayoutTests/fast/forms/select/select-focus-ring-is-not-clipped.html:
```diff
@@ -0,0 +1,37 @@
+        margin: 0;
+        padding: 0;
+        padding-top: 5px;
+	 }
```
Ditto.

pxlcoder commented on LayoutTests/fast/forms/select/select-focus-ring-is-not-clipped-expected-mismatch.html:
```diff
@@ -0,0 +1,27 @@
+<!DOCTYPE html>
+<html>
```
Does the test do what you expect on both macOS and iOS? You might want to add a meta viewport tag.

lilyspiniolas replied:
Added. The page looks as expected on iOS.

pxlcoder commented on LayoutTests/fast/forms/select/select-focus-ring-is-not-clipped.html:
```diff
@@ -0,0 +1,37 @@
+    }
+
+    async function runTest() {
+        await UIHelper.activateAt(23, 10);
```
If this is about activating the `select`, there's `activateElement`.

pxlcoder reviewed (approved):
r=mews

lilyspiniolas committed:
[Form Control Refresh] Focus ring clipped for native <select> buttons
https://bugs.webkit.org/show_bug.cgi?id=295256
rdar://154729412

Reviewed by Aditya Keerthi.

After 296129@main, native select buttons no longer had their border
radius reset, causing bleed avoidance to apply which resulted in the
focus ring getting clipped.

With these changes, bleed avoidance (except when handled directly by
the theme) no longer applies to any painted control.

* LayoutTests/fast/forms/select/select-focus-ring-is-not-clipped-expected-mismatch.html: Added.
* LayoutTests/fast/forms/select/select-focus-ring-is-not-clipped.html: Added.
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::determineBleedAvoidance const):
* Source/WebCore/rendering/RenderTheme.h:
(WebCore::RenderTheme::mayNeedBleedAvoidance const):
* Source/WebCore/rendering/cocoa/RenderThemeCocoa.h:
* Source/WebCore/rendering/cocoa/RenderThemeCocoa.mm:
(WebCore::RenderThemeCocoa::mayNeedBleedAvoidance const):

Canonical link: https://commits.webkit.org/296951@main
